dvc/state.py

```diff
@@ -443,36 +443,37 @@ def save_link(self, path_info):
         )
         self._execute(cmd, (relative_path, self._to_sqlite(inode), mtime))
 
-    def remove_unused_links(self, used):
-        """Removes all saved links except the ones that are used.
+    def get_unused_links(self, used):
+        """Returns all saved links except the ones that are used.
 
         Args:
-            used (list): list of used links that should not be removed.
+            used (list): list of used links
         """
         unused = []
 
         self._execute("SELECT * FROM {}".format(self.LINK_STATE_TABLE))
         for row in self.cursor:
-            relpath, inode, mtime = row
+            relative_path, inode, mtime = row
             inode = self._from_sqlite(inode)
-            path = os.path.join(self.root_dir, relpath)
+            path = os.path.join(self.root_dir, relative_path)
 
-            if path in used:
-                continue
-
-            if not os.path.exists(path):
+            if path in used or not os.path.exists(path):
                 continue
 
             actual_inode = get_inode(path)
             actual_mtime, _ = get_mtime_and_size(path, self.repo.tree)
 
-            if inode == actual_inode and mtime == actual_mtime:
-                logger.debug("Removing '{}' as unused link.", path)
-                remove(path)
-                unused.append(relpath)
+            if (inode, mtime) == (actual_inode, actual_mtime):
+                unused.append(relative_path)
+
+        return unused
+
+    def remove_links(self, links):
+        for link in links:
+            remove(os.path.join(self.root_dir, link))
```

Comment:
A more natural refactor would split old `remove_unused_links` into `get_unused_links` and `remove_link(s)`(as a reverse `save_link`). So we would `get_unused_links`, remove them in `checkout` and then `state.remove_links()` what we've deleted. That way the context would be properly spread across these parts of code.